dev-tools/xtask/src/virtual_hardware.rs

```diff
@@ -0,0 +1,834 @@
+    let mut cmd = Command::new(PFEXEC);
+    cmd.args([
+        ZLOGIN,
+        "sidecar_softnpu",
+        "/softnpu/scadm",
+        "--server",
+        "/softnpu/server",
+        "--client",
+        "/softnpu/client",
+        "standalone",
+    ]);
+    for arg in &args {
+        cmd.arg(arg);
+    }
+    Ok(execute(cmd)?)
+}
+
+fn default_gateway_ip() -> Result<String> {
+    let mut cmd = Command::new(NETSTAT);
+    cmd.args(["-rn", "-f", "inet"]);
+    let output = execute(cmd)?;
+
+    String::from_utf8(output.stdout)
+        .context("Invalid netstat output")?
+        .lines()
+        .find_map(|line| {
+            let mut columns = line.trim().split_whitespace();
+            let dst = columns.next()?;
+            let gateway = columns.next()?;
+
+            if dst == "default" {
+                return Some(gateway.to_owned());
+            }
+            None
+        })
+        .ok_or_else(|| anyhow!("No default gateway found"))
+}
+
+fn get_gateway_mac(
+    gateway_mac: Option<String>,
+    gateway_ip: &str,
+) -> Result<MacAddr> {
+    match gateway_mac {
+        Some(mac) => Ok(MacAddr::from_str(&mac)?),
+        None => {
+            let retries = 3;
+            for i in 0..=retries {
+                println!("Pinging {gateway_ip} and sleeping ({i} / {retries})");
+                let mut cmd = Command::new(PING);
+                cmd.arg(&gateway_ip);
+                execute(cmd)?;
+                std::thread::sleep(std::time::Duration::from_secs(1));
+            }
+
+            let mut cmd = Command::new(ARP);
+            cmd.arg("-an");
+            let output = execute(cmd)?;
+
+            let mac = String::from_utf8(output.stdout)
+                .context("Invalid arp output")?
+                .lines()
+                .find_map(|line| {
+                    let mut columns = line.trim().split_whitespace().skip(1);
+                    let ip = columns.next()?;
+                    let mac = columns.last()?;
+                    if ip == gateway_ip {
+                        return Some(mac.to_string());
+                    }
+                    None
+                })
+                .ok_or_else(|| anyhow!("No gateway MAC found"))?;
+            Ok(MacAddr::from_str(&mac)?)
+        }
+    }
+}
+
+/// This is a subset of omicron-sled-agent's "config/Config" structure.
+///
+/// We don't depend on it directly to avoid rebuilding whenever the
+/// Sled Agent changes, though it's important for us to stay in sync
+/// to parse these fields correctly.
+#[derive(Clone, Debug, Deserialize)]
+struct SledAgentConfig {
+    /// Optional list of virtual devices to be used as "discovered disks".
+    pub vdevs: Option<Vec<Utf8PathBuf>>,
+}
+
+impl SledAgentConfig {
+    fn read(path: &Utf8Path) -> Result<Self> {
+        let config = std::fs::read_to_string(path)?;
+        Ok(toml::from_str(&config)
+            .context("Could not parse sled agent config as toml")?)
+    }
+}
```

Comment:
Using this structure instead of "sed" resolves https://github.com/oxidecomputer/omicron/issues/5401